sys/aarch64/pmap.c

```diff
@@ -496,25 +496,26 @@ void pmap_set_modified(vm_page_t *pg) {
 
 int pmap_emulate_bits(pmap_t *pmap, vaddr_t va, vm_prot_t prot) {
   paddr_t pa;
+  vm_page_t *pg = NULL;
+  pte_t pte = 0;
 
   WITH_MTX_LOCK (&pmap->mtx) {
     if (!pmap_extract_nolock(pmap, va, &pa))
       return EFAULT;
 
-    pte_t pte = *pmap_lookup_pte(pmap, va);
+    pte = *pmap_lookup_pte(pmap, va);
+  }
+  pg = vm_page_find(pa);
+  assert(pg != NULL);
 
-    if ((prot & VM_PROT_READ) && !(pte & ATTR_SW_READ))
-      return EACCES;
+  if ((prot & VM_PROT_READ) && !(pte & ATTR_SW_READ))
+    return EACCES;
 
-    if ((prot & VM_PROT_WRITE) && !(pte & ATTR_SW_WRITE))
-      return EACCES;
+  if ((prot & VM_PROT_WRITE) && !(pte & ATTR_SW_WRITE))
+    return EACCES;
 
-    if ((prot & VM_PROT_EXEC) && (pte & ATTR_SW_NOEXEC))
-      return EACCES;
-  }
-
-  vm_page_t *pg = vm_page_find(pa);
-  assert(pg != NULL);
+  if ((prot & VM_PROT_EXEC) && (pte & ATTR_SW_NOEXEC))
+    return EACCES;
 
```

Comment:
I don't understand that change. Why did you move `vm_page_find` before access checking? I don't see any advantages.
Please don't set `pte` and `pg` at the beginning of that function - it's confusing. You always use only values returned by `pmap_lookup_pte` and `vm_page_find` so _default_ 0 is not good here.

Reply:
I moved `vm_page_find` before access checking because in copy-on-write was the case where we checked access and raised segfault after that. However because the page we checked access for, belonged to the backing object, we shouldn't return `EACCES` but` EFAULT` to make sure `vm_page_fault` will be executed. That's the reason for change.

Reply:
But here you don't check if page belongs to backing object.

Reply:
I forgot to but I fixed it and now I check it.